### PR TITLE
AutocompleteRemoveFragment: Get domain selection from view before view is getting destroyed. (#1838)

### PR DESCRIPTION
**Pull request by pocmo**

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/autocomplete/AutocompleteRemoveFragment.kt [hunk 1]
```diff
@@ -28,14 +29,16 @@ class AutocompleteRemoveFragment : AutocompleteListFragment() {
...
     private fun removeSelectedDomains(context: Context) {
-        launch(CommonPool) {
-            val domains = (domainList.adapter as DomainListAdapter).selection()
-            if (!domains.isEmpty()) {
-                CustomAutocomplete.removeDomains(context, domains)
+        val domains = (domainList.adapter as DomainListAdapter).selection()
+        if (domains.isNotEmpty()) {
+            launch(UI) {
+                async {
+                    CustomAutocomplete.removeDomains(context, domains)
```
Comment (liuche) [inline app/src/main/java/org/mozilla/focus/autocomplete/AutocompleteRemoveFragment.kt]: I see, this async/await is capturing the context before popping the backstack, and then finishing it on a background thread.

**Comment by liuche** [comment]
Note: this needs to be rebased because of conflicts

**Comment by mcomella** [comment]
Merge conflict is because:
```kotlin
CustomAutocomplete.removeDomains(context, domains)
TelemetryWrapper.removeAutocompleteDomainsEvent(domains.size) // <- this line added
```

Working on merging...

**Comment by mcomella** [comment]
Ahh, merge conflict is because the first patch does not contain that line ^ even though the aggregation of all the patches contains it. I think I can fix this by squashing all the commits and then rebasing.

@pocmo Why do all these merge commits exist?
screen shot 2017-12-01 at 16 08 27

fwiw, if they weren't there, and catching up to master occurred with `git rebase` instead, this conflict wouldn't have happened.

**Comment by mcomella** [comment]
I can't seem to figure out how to get rid of the merge commits with `git rebase -i` (I was hoping to squash): I'm guessing Sebastian manually merged to master by hand so git doesn't know how to manage that.

I think I'll just rewrite the commit using absolute changes, from the "Files changed" tab.

@pocmo, how would you fix this for future reference? I guess one benefit of the `merge` commits is that you create an absolute history of what has changed rather than my more destructive `git rebase up/master && git push -f`.

**Comment by mcomella** [comment]
> I can't seem to figure out how to get rid of the merge commits with git rebase -i (I was hoping to squash): I'm guessing Sebastian manually merged to master by hand so git doesn't know how to manage that.

Nevermind – I ended up doing `git rebase up/master`, fixing the first merge error by hand, and git was smart enough to remove the other merge commits. It seems to defeat the point of having the merge commits though if I need to redo the merges by hand. I'm quite curious to know how Sebastian solves this! :)

**Comment by mcomella** [comment]
Merged with https://github.com/mozilla-mobile/focus-android/commit/b505ea7fc5b65cd410b7bf4253636fad3232ae2e

File: app/src/main/java/org/mozilla/focus/autocomplete/AutocompleteRemoveFragment.kt [hunk 2]
```diff
@@ -29,16 +30,18 @@ class AutocompleteRemoveFragment : AutocompleteListFragment() {
...
     private fun removeSelectedDomains(context: Context) {
-        launch(CommonPool) {
-            val domains = (domainList.adapter as DomainListAdapter).selection()
-            if (!domains.isEmpty()) {
-                CustomAutocomplete.removeDomains(context, domains)
-
-                TelemetryWrapper.removeAutocompleteDomainsEvent(domains.size)
+        val domains = (domainList.adapter as DomainListAdapter).selection()
+        if (domains.isNotEmpty()) {
+            launch(UI) {
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/focus/autocomplete/AutocompleteRemoveFragment.kt]: I think the coroutine code can be simplified. Currently, it's:
```kotlin
            launch(UI) {
                async {
                    CustomAutocomplete.removeDomains(context, domains)

                    TelemetryWrapper.removeAutocompleteDomainsEvent(domains.size)
                }.await()

                fragmentManager.popBackStack()
            }
```

Functionally equivalent, I think this is more readable:
```kotlin
            launch {
                CustomAutocomplete.removeDomains(context, domains)
                TelemetryWrapper.removeAutocompleteDomainsEvent(domains.size)
                launch(UI) {
                    fragmentManager.popBackStack()
                }
            }
```

But it seems `fragmentManager.popBackStack()` could be thread-safe (it queues the popping and returns) so I believe:
```kotlin
            launch {
                CustomAutocomplete.removeDomains(context, domains)
                TelemetryWrapper.removeAutocompleteDomainsEvent(domains.size)
                fragmentManager.popBackStack()
            }
```

Should also work.

**Pull request closed without merging**